resolwe/flow/models/data.py

```diff
@@ -430,14 +441,28 @@ def save(self, render_name=False, *args, **kwargs):  # pylint: disable=keyword-a
                                 test_required=False)
 
         with transaction.atomic():
-            super().save(*args, **kwargs)
+            self._save(*args, **kwargs)
 
             # We can only save dependencies after the data object has been saved. This
             # is why a transaction block is needed and the save method must be called first.
             if create:
                 self.save_dependencies(self.input, self.process.input_schema)  # pylint: disable=no-member
                 self.create_entity()
 
+    @transaction.atomic
+    def _save(self, *args, **kwargs):
```

Comment:
Why have you added a separate method for this? There is already a ton of stuff in existing `save`.

Reply:
This `_save` is the absolute minimum _save_. If you, for example, want to implement duplicate function, one would call `_save` inside it, because you don't need validation and whatnot, but you still need to create `DataLocation` instance.

Reply:
This function is too specific to be used in duplication function as it sets data location to data it, which is only true when new data object is created.

Reply:
I agree that this split is useful, I just wanted to say that it doesn't support all use-cases, because it sets `location_id=str(self.pk)`.
And i am in favour of a more descriptive name instead of `_save` - maybe `perform_save`, to follow DRF's practice.

Reply:
>And i am in favour of a more descriptive name instead of _save - maybe perform_save, to follow DRF's practice.

That makes it a "public" function. Does it need to be decorated with `@transaction.atomic`?

Reply:
Good question :).

If a Data object is created over the API, there are at least three transactions  ([here](https://github.com/genialis/resolwe/blob/master/resolwe/flow/views/data.py#L174), [here](https://github.com/genialis/resolwe/blob/master/resolwe/flow/models/data.py#L432) and this one). And there are still lines like [this one](https://github.com/genialis/resolwe/blob/master/resolwe/flow/models/data.py#L398) that should be in s transaction, but are not always (e.g. when executor updates the output).

A good solution nay be to name the method `_perform_create` and remove the transaction.